### PR TITLE
enhance exception handling and change to build to gradle

### PR DESCRIPTION
**Pull request by DC2-DanielKrueger**

### CODE REVIEW

**Review by Florian-Limpoeck — APPROVED** [review]

File: .gitignore [hunk 1]
```diff
@@ -1,10 +1,15 @@
...
 
-dependency-reduced-pom.xml
+.idea/*
+!.idea/codeStyles
+!.idea/inspectionProfiles
+!.idea/runConfigurations
+.java-version
+.DS_Store
+
+/hivemq4-prometheus-extension.iml
```
Comment (SgtSilvio) [inline .gitignore:15]: please use the same .gitignore as in other projects
```
# Gradle
.gradle
build/

# IntelliJ
out/
*.iml
.idea/*
!.idea/codeStyles
!.idea/inspectionProfiles
!.idea/runConfigurations

.java-version
.DS_Store
```

File: build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+val prepareAsciidocTask = tasks.register<Sync>("prepareAsciidoc") {
+    from("README.adoc").into(buildDir.resolve("tmp/asciidoc"))
+}
+tasks.asciidoctor {
+    dependsOn(prepareAsciidocTask)
+    sourceDir(prepareAsciidocTask.get().outputs.files.asPath)
+}
+
+
+
```
Comment (SgtSilvio) [inline build.gradle.kts:33]: please remove unnecessary whitespaces

File: build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    testImplementation("junit:junit:${property("junit.version")}")
+    testImplementation("com.hivemq:hivemq-testcontainer-junit4:${property("hivemq-testcontainer.version")}")
+    testImplementation("commons-io:commons-io:${property("commons-io.version")}")
+}
+
+
+
+
+
+
```
Comment (SgtSilvio) [inline build.gradle.kts:52]: please remove unnecessary whitespaces

File: settings.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+pluginManagement {
+    plugins {
+        id("com.hivemq.extension") version "${extra["plugin.hivemq-extension.version"]}"
+        id("org.asciidoctor.jvm.convert") version "${extra["plugin.asciidoctor.version"]}"
+    }
+}
+
+
+
```
Comment (SgtSilvio) [inline settings.gradle.kts:9]: please remove unnecessary whitespaces

File: gradle.properties [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+# main dependencies
+#
+hivemq-extension-sdk.version=4.0.0
+simpleclient-dropwizard.version=0.9.0
+simpleclient-servlet.version = 0.9.0
+simpleclient-common.version= 0.9.0
+jetty-server.version= 10.0.0.beta2
+jetty-servlet.version= 10.0.0.beta2
+owner.version=1.0.10
+mockito-all.version=1.10.19
```
Comment (SgtSilvio) [inline gradle.properties:12]: please separate test and main dependencies

File: build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+dependencies {
+    implementation("io.prometheus:simpleclient_dropwizard:${property("simpleclient-dropwizard.version")}")
+    implementation("io.prometheus:simpleclient_servlet:${property("simpleclient-servlet.version")}")
+    implementation("io.prometheus:simpleclient_common:${property("simpleclient-common.version")}")
+    implementation("org.eclipse.jetty:jetty-server:${property("jetty-server.version")}")
+    implementation("org.eclipse.jetty:jetty-servlet:${property("jetty-servlet.version")}")
+    implementation("org.aeonbits.owner:owner:${property("owner.version")}")
+
+    testImplementation("org.mockito:mockito-all:${property("mockito-all.version")}")
+    testImplementation("junit:junit:${property("junit.version")}")
```
Comment (SgtSilvio) [inline build.gradle.kts:43]: I like to have junit as the first test dependency

File: gradle.properties [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+junit.version=4.13
+commons-io.version=2.4
+hivemq-testcontainer.version=1.2.0
```
Comment (SgtSilvio) [inline gradle.properties:15]: please use the same order as you defined the dependencies also for the versions

File: gradle.properties (hunk 5 above)
Comment (SgtSilvio) [inline gradle.properties:12]:
```suggestion
mockito.version=1.10.19
```

File: .gitignore (hunk 1 above)
Comment (DC2-DanielKrueger) [reply .gitignore:15]: done

File: build.gradle.kts (hunk 2 above)
Comment (DC2-DanielKrueger) [reply build.gradle.kts:33]: done

File: settings.gradle.kts (hunk 4 above)
Comment (DC2-DanielKrueger) [reply settings.gradle.kts:9]: done

File: build.gradle.kts (hunk 6 above)
Comment (DC2-DanielKrueger) [reply build.gradle.kts:43]: For Real? One could say, it´s a bit picky, but ur wish is my command.

File: build.gradle.kts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    priority = 0
+    startPriority = 0
+    sdkVersion = "${property("hivemq-extension-sdk.version")}"
+}
+
+tasks.hivemqExtensionResources {
+    from("LICENSE")
+    from("README.adoc") { rename { "README.txt" } }
+    from(tasks.asciidoctor)
+    from("src/main/resources/prometheusConfiguration.properties")
```
Comment (SgtSilvio) [inline build.gradle.kts:21]: can you instead move the properties file to the src/hivemq-extension/ folder and remove this line

File: gradle.properties (hunk 7 above)
Comment (DC2-DanielKrueger) [reply gradle.properties:15]: done

File: gradle.properties (hunk 5 above)
Comment (DC2-DanielKrueger) [reply gradle.properties:12]: done

File: build.gradle.kts (hunk 8 above)
Comment (DC2-DanielKrueger) [reply build.gradle.kts:21]: done

File: build.gradle.kts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    id("org.asciidoctor.jvm.convert")
+}
+
+group = "com.hivemq"
+description = "Prometheus Monitoring Extension"
+
+hivemqExtension {
+    name = "Prometheus Monitoring Extension"
+    author = "HiveMQ"
+    priority = 0
```
Comment (SgtSilvio) [inline build.gradle.kts:12]:
```suggestion
    priority = 1000
```
Should not be changed, before it was 1000

File: src/main/java/com/hivemq/extensions/prometheus/export/MonitoredMetricServlet.java [hunk 10]
```diff
@@ -55,12 +53,15 @@
 
 
     @Override
-    protected void doGet(final HttpServletRequest req, final HttpServletResponse resp) throws ServletException, IOException {
-        log.debug("Received HTTP-Get-Request from Prometheus to scrape metrics from HiveMQ.", req.toString());
+    protected void doGet(final HttpServletRequest req, final HttpServletResponse resp) {
+        log.debug("Received HTTP-Get-Request from Prometheus to scrape metrics from HiveMQ: {}", req.toString());
         final Timer.Context context = responses.time();
-        super.doGet(req, resp);
+       try{
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/extensions/prometheus/export/MonitoredMetricServlet.java:59]: please format

File: build.gradle.kts (hunk 6 above)
Comment (SgtSilvio) [reply build.gradle.kts:40]: you are correct that this is minor, but everything that makes dependencies easier to maintain is a good thing

File: gradle.properties [hunk 11]
```diff
@@ -0,0 +1,24 @@
+version=4.0.4
+#
+# main dependencies
+#
+hivemq-extension-sdk.version=4.0.0
+simpleclient-dropwizard.version=0.9.0
+simpleclient-servlet.version = 0.9.0
+simpleclient-common.version= 0.9.0
```
Comment (SgtSilvio) [inline gradle.properties:8]: please use one version
```suggestion
prometheus-simpleclient.version=0.9.0
```

File: gradle.properties [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+jetty-server.version= 10.0.0.beta2
+jetty-servlet.version= 10.0.0.beta2
```
Comment (SgtSilvio) [inline gradle.properties:10]:
```suggestion
jetty.version=10.0.0.beta2
```

File: build.gradle.kts [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    startPriority = 0
+    sdkVersion = "${property("hivemq-extension-sdk.version")}"
+}
+
+tasks.hivemqExtensionResources {
+    from("LICENSE")
```
Comment (SgtSilvio) [inline build.gradle.kts:18]:
```suggestion
    from("LICENSE.txt")
```

File: build.gradle.kts [hunk 14]
```diff
@@ -0,0 +1,43 @@
+plugins {
+    id("com.hivemq.extension")
+    id("org.asciidoctor.jvm.convert")
+}
+
+group = "com.hivemq"
+description = "Prometheus Monitoring Extension"
```
Comment (SgtSilvio) [inline build.gradle.kts:7]:
```suggestion
description = "HiveMQ extension to utilize Prometheus for monitoring"
```

File: build.gradle.kts [hunk 15]
```diff
@@ -0,0 +1,43 @@
+plugins {
+    id("com.hivemq.extension")
+    id("org.asciidoctor.jvm.convert")
+}
+
+group = "com.hivemq"
```
Comment (SgtSilvio) [inline build.gradle.kts:6]:
```suggestion
group = "com.hivemq.extensions"
```

File: gradle.properties (hunk 12 above)
Comment (SgtSilvio) [inline gradle.properties:10]: why do we now depend on a beta version?

**Review by SgtSilvio — CHANGES_REQUESTED** [review]
see comments

File: pom.xml [hunk 16]
```diff
@@ -1,245 +0,0 @@
...
-                                    <exclude>com.google.guava:*</exclude>
-                                    <exclude>org.slf4j:*</exclude>
-                                    <exclude>ch.qos.logback:*</exclude>
-                                    <exclude>javax.ws.rs:*</exclude>
-                                </excludes>
-                            </artifactSet>
-                            <relocations>
-                                <relocation>
-                                    <pattern>javax.servlet</pattern>
-                                    <shadedPattern>com.hivemq.shaded.javax.servlet</shadedPattern>
```
Comment (SgtSilvio) [inline pom.xml:131]: with the gradle build we do not shade the javax.servlet package anymore, I don't know if this is needed at all.

File: build.gradle.kts (hunk 9 above)
Comment (DC2-DanielKrueger) [reply build.gradle.kts:12]: done

File: src/main/java/com/hivemq/extensions/prometheus/export/MonitoredMetricServlet.java (hunk 10 above)
Comment (DC2-DanielKrueger) [reply src/main/java/com/hivemq/extensions/prometheus/export/MonitoredMetricServlet.java:59]: done

File: gradle.properties (hunk 12 above)
Comment (DC2-DanielKrueger) [reply gradle.properties:10]: yeah moved back.

File: pom.xml (hunk 16 above)
Comment (DC2-DanielKrueger) [reply pom.xml:131]: i dont think we need it at all

File: gradle.properties [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+junit.version=4.13
+commons-io.version=2.4
+hivemq-testcontainer.version=1.2.0
+#
+# plugins
+#
+plugin.hivemq-extension.version=1.0.0
+plugin.asciidoctor.version=3.2.0
+
+
```
[the remark this answers is not in the document]
Comment (DC2-DanielKrueger) [reply gradle.properties:24]: done

File: build.gradle.kts (hunk 8 above)
Comment (SgtSilvio) [reply build.gradle.kts:21]: This is still here, it should be removed

File: src/main/java/com/hivemq/extensions/prometheus/export/MonitoredMetricServlet.java [hunk 18]
```diff
@@ -55,12 +53,15 @@
...
-    protected void doGet(final HttpServletRequest req, final HttpServletResponse resp) throws ServletException, IOException {
-        log.debug("Received HTTP-Get-Request from Prometheus to scrape metrics from HiveMQ.", req.toString());
+    protected void doGet(final HttpServletRequest req, final HttpServletResponse resp) {
+        log.debug("Received HTTP-Get-Request from Prometheus to scrape metrics from HiveMQ: {}", req.toString());
         final Timer.Context context = responses.time();
-        super.doGet(req, resp);
+        try {
+            super.doGet(req, resp);
+        } catch (Exception e) {
+            log.warn("Exception occurred while collection of metrics and creation of Prometheus Servlet: {}.", e.getClass().getSimpleName());
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/extensions/prometheus/export/MonitoredMetricServlet.java:62]:
```suggestion
            log.warn("Exception occurred while collecting metrics and creating of Prometheus response: {}.", e.getClass().getSimpleName());
```

File: src/main/java/com/hivemq/extensions/prometheus/export/MonitoredMetricServlet.java [hunk 19]
```diff
@@ -55,12 +53,15 @@
...
+            log.debug("Original exception was :", e);
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/extensions/prometheus/export/MonitoredMetricServlet.java:63]:
```suggestion
            log.debug("Original exception was: ", e);
```

File: src/main/java/com/hivemq/extensions/prometheus/export/MonitoredMetricServlet.java [hunk 20]
```diff
@@ -55,12 +53,15 @@
 
 
     @Override
-    protected void doGet(final HttpServletRequest req, final HttpServletResponse resp) throws ServletException, IOException {
-        log.debug("Received HTTP-Get-Request from Prometheus to scrape metrics from HiveMQ.", req.toString());
+    protected void doGet(final HttpServletRequest req, final HttpServletResponse resp) {
+        log.debug("Received HTTP-Get-Request from Prometheus to scrape metrics from HiveMQ: {}", req.toString());
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/extensions/prometheus/export/MonitoredMetricServlet.java:57]:
```suggestion
        log.debug("Received HTTP-Get-Request from Prometheus to scrape metrics from HiveMQ: {}", req);
```

File: src/main/java/com/hivemq/extensions/prometheus/export/MonitoredMetricServlet.java (hunk 18 above)
Comment (DC2-DanielKrueger) [reply src/main/java/com/hivemq/extensions/prometheus/export/MonitoredMetricServlet.java:62]: done

File: src/main/java/com/hivemq/extensions/prometheus/export/MonitoredMetricServlet.java (hunk 19 above)
Comment (DC2-DanielKrueger) [reply src/main/java/com/hivemq/extensions/prometheus/export/MonitoredMetricServlet.java:63]: done

File: src/main/java/com/hivemq/extensions/prometheus/export/MonitoredMetricServlet.java (hunk 20 above)
Comment (DC2-DanielKrueger) [reply src/main/java/com/hivemq/extensions/prometheus/export/MonitoredMetricServlet.java:57]: done

File: build.gradle.kts (hunk 14 above)
Comment (DC2-DanielKrueger) [reply build.gradle.kts]: done

File: src/main/java/com/hivemq/extensions/prometheus/export/MonitoredMetricServlet.java [hunk 21]
```diff
@@ -55,12 +53,15 @@
...
-        log.debug("Received HTTP-Get-Request from Prometheus to scrape metrics from HiveMQ.", req.toString());
+    protected void doGet(final HttpServletRequest req, final HttpServletResponse resp) {
+        log.debug("Received HTTP-Get-Request from Prometheus to scrape metrics from HiveMQ: {}", req);
         final Timer.Context context = responses.time();
-        super.doGet(req, resp);
+        try {
+            super.doGet(req, resp);
+        } catch (Exception e) {
+            log.warn("Exception occurred while collecting metrics and creating of Prometheus response: {}.", e.getClass().getSimpleName());
+            log.debug("Original exception was:", e);
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/extensions/prometheus/export/MonitoredMetricServlet.java:63]:
```suggestion
            log.debug("Original exception was: ", e);
```

File: gradle.properties [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+#
+hivemq-extension-sdk.version=4.0.0
+simpleclient.version=0.9.0
+jetty.version=9.4.27.v20200227
+owner.version=1.0.10
+#
+# test dependencies
+#
+junit.version=4.13
+mockito.version=1.10.19
```
Comment (SgtSilvio) [inline gradle.properties:13]:
```suggestion
junit.version=4.13
mockito.version=1.10.19
test
```

**Review by SgtSilvio — APPROVED** [review]

File: build.gradle.kts [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+group = "com.hivemq.extensions"
+description = "HiveMQ extension to utilize Prometheus for monitoring"
+
+hivemqExtension {
+    name = "Prometheus Monitoring Extension"
+    author = "HiveMQ"
+    priority = 1000
+    startPriority = 0
```
Comment (SgtSilvio) [inline build.gradle.kts:14]: @DC2-DanielKrueger the default start priority is 1000, so we should not change it

Comment (SgtSilvio) [inline build.gradle.kts:14]:
```suggestion
    startPriority = 1000
```

**Pull request merged**